File: python/54_poker_hands.py
from enum import Enum
# using enums for custom comparison of types

class HandRanking(Enum):
    # hand rankings ordered by score, lowest to highest
    HIGH_CARD = 1
    PAIR = 2
    TWO_PAIR = 3
    THREE_OF_A_KIND = 4
    STRAIGHT = 5
    FLUSH = 6
    FULL_HOUSE = 7
    FOUR_OF_A_KIND = 8
    STRAIGHT_FLUSH = 9
    ROYAL_FLUSH = 10
    
    # compare hand rankings
    def __lt__(h1, h2):
        return h1.value < h2.value

def poker_hands(hands: list) -> int:
    
    count: int = 0
    
    for hand in hands:
        cards = hand.split()
        hand1 = cards[:5]
        hand2 = cards[5:]
        
        score1 = evaluate(hand1)
        score2 = evaluate(hand2)
        
        if score1["ranking"] > score2["ranking"] or (score1["ranking"] == score2["ranking"] and score1["kicker"] > score2["kicker"]):
            count += 1
        
    return count
    
    
def evaluate(hand):
    values = ["2", "3", "4", "5", "6", "7", "8", "9", "T", "J", "Q", "K", "A"]
    suits = ["C", "D", "H", "S"]
    
    val_occs_dict = {value: 0 for value in values}
    val_occs_arr = [0] * 13
    suit_occs = [0] * 4
    
    for card in hand:
        val_occs_dict[card[0]] += 1
        val_occs_arr[values.index(card[0])] += 1
        suit_occs[suits.index(card[1])] += 1
    
    # include kicker entry in res as tiebreaker for certain rankings
    streaks = "".join(str(val) for val in val_occs_arr)
    max_streak = max(len(streak) for streak in streaks.split("0"))
    is_straight = max_streak == 5 or ("1111" in streaks and streaks.index("1111") == 0 and val_occs_dict["A"] == 1) # five-high straight
    is_flush = 5 in suit_occs
    
    res = {}
    
    # Royal Flush / Straight Flush
    if is_straight and is_flush:
        if all(val_occs_dict[val] == 1 for val in values[8:]):
            print("royal flush")
            res["ranking"] = HandRanking.ROYAL_FLUSH
            return res
        else:
            print("straight flush")
            res["ranking"] = HandRanking.STRAIGHT_FLUSH
        res["kicker"] = max(idx for idx, val in enumerate(val_occs_arr[:12]) if val == 1)
        return res
    
    # Flush
    if is_flush:
        res["ranking"] = HandRanking.FLUSH
        return res
    
    # Straight
    if is_straight:
        # print("straight")
        res["ranking"] = HandRanking.STRAIGHT
        res["kicker"] = max(idx for idx, val in enumerate(val_occs_arr[:12]) if val == 1)
        return res
    
    # Four of a Kind
    if 4 in val_occs_dict.values():
        res["ranking"] = HandRanking.FOUR_OF_A_KIND
        res["kicker"] = val_occs_arr.index(4)
        return res
    
    # Full House
    if all(num in val_occs_dict.values() for num in [2, 3]):
        res["ranking"] = HandRanking.FULL_HOUSE
        res["kicker"] = val_occs_arr.index(3)
        return res
    
    # Three of a Kind
    if 3 in val_occs_dict.values():
        res["ranking"] = HandRanking.THREE_OF_A_KIND
        res["kicker"] = val_occs_arr.index(3)
        return res
    
    # Two Pair / One Pair
    if 2 in val_occs_dict.values():
        res["ranking"] = HandRanking.TWO_PAIR if list(val_occs_dict.values()).count(2) == 2 else HandRanking.PAIR
        res["kicker"] = max(idx for idx, val in enumerate(val_occs_arr) if val == 2)
        return res
    
    # High Card
    res["ranking"] = HandRanking.HIGH_CARD
    res["kicker"] = max(idx for idx, val in enumerate(val_occs_arr) if val == 1)
    return res

if __name__ == "__main__":
    f = open("54_poker.txt", 'r')
    lines = f.readlines()
    hands = [line.strip() for line in lines]
    f.close()
    
    print(poker_hands(hands))

File: cpp/54_poker_hands.cpp
#include <iostream>
#include <fstream>
#include <vector>
#include <string>
#include <cstring>
#include <array>
#include <algorithm>

std::array<int, 2> evaluate(std::string hand[5]){
    int values[13] = {};
    int suits[4] = {};
    
    for (int i = 0; i < 5; i++){
        std::string card = hand[i];
        switch(card[0]){
            case 'T': values[8]++; break;
            case 'J': values[9]++; break;
            case 'Q': values[10]++; break;
            case 'K': values[11]++; break;
            case 'A': values[12]++; break;
            default: values[card[0] - '0' - 2]++;
        }
        switch(card[1]){
            case 'C': suits[0]++; break;
            case 'D': suits[1]++; break;
            case 'H': suits[2]++; break;
            case 'S': suits[3]++; break;
        }
    }
    std::string streaks = "";
    for (int i = 0; i < 13; i++){
        streaks += std::to_string(values[i]);
    }
    bool straight = (streaks.find("11111") != std::string::npos) or (streaks.find("1111") == 0 and values[12] == 1);
    bool flush = std::count(std::begin(suits), std::end(suits), 5) == 1;
    if (straight and flush){
        // royal flush
        if (values[8] and values[12]) return {9, 0};
        // five-high straight flush
        if (values[3] and values[12]) return {8, 3};
        int kicker = 0;
        for (int i = 12; i >= 0; i--){
            if (values[i] != 0){
                kicker = i; 
                break;
            }
        }
        return {8, kicker};
    }
    if (flush){
        int kicker = 0;
        for (int i = 12; i >= 0; i--){
            if (values[i] != 0){
                kicker = i; 
                break;
            }
        }
        return {5, kicker};
    }
    if (straight){
        if (values[3] and values[12]) return {4, 3};
        int kicker = 0;
        for (int i = 12; i >= 0; i--){
            if (values[i] != 0){
                kicker = i; 
                break;
            }
        }
        return {4, kicker};
    }
    
    bool four = std::find(std::begin(values), std::end(values), 4) != std::end(values);
    if (four){
        int kicker = 0;
        for (int i = 12; i >= 0; i--){
            if (values[i] == 4){
                kicker = i; 
                break;
            }
        }
        return {7, kicker};
    }
    
    bool three = std::count(std::begin(values), std::end(values), 3) == 1;
    bool two_pair = std::count(std::begin(values), std::end(values), 2) == 2;
    bool pair = std::count(std::begin(values), std::end(values), 2) == 1;
    
    if (three and pair){
        int kicker = 0;
        for (int i = 12; i >= 0; i--){
            if (values[i] == 3){
                kicker = i; 
                break;
            }
        }
        return {6, kicker};
    }
    if (three){
        int kicker = 0;
        for (int i = 12; i >= 0; i--){
            if (values[i] == 3){
                kicker = i; 
                break;
            }
        }
        return {3, kicker};
    }
    if (two_pair){
        int kicker = 0;
        for (int i = 12; i >= 0; i--){
            if (values[i] == 2){
                kicker = i; 
                break;
            }
        }
        return {2, kicker};
    }
    if (pair){
        int kicker = 0;
        for (int i = 12; i >= 0; i--){
            if (values[i] == 2){
                kicker = i; 
                break;
            }
        }
        return {1, kicker};
    }
    int kicker = 0;
    for (int i = 12; i >= 0; i--){
        if (values[i] != 0){
            kicker = i; 
            break;
        }
    }
    return {0, kicker};
}

int poker_hands(std::vector<std::string> hands){
    int count = 0;
    for (std::string hand : hands){
        std::string h1[5] = {};
        std::string h2[5] = {};
        for (int i = 0; i < 5; i++){
            h1[i] = hand.substr(3*i, 2);
            h2[i] = hand.substr(3*i + 15, 2);
        }
        std::array<int, 2> score1 = evaluate(h1);
        std::array<int, 2> score2 = evaluate(h2);
        if (score1[0] > score2[0] || (score1[0] == score2[0] && score1[1] > score2[1])){
            count++;
        }
    }
    return count;
}

int main(){
    std::ifstream file("/uploads/54_poker.txt");
    std::string s;
    std::vector<std::string> hands;
    
    while (std::getline(file, s)){
        hands.push_back(s);
    }
    file.close();
    //std::string test[5] = {"AD", "KD", "JD", "QD", "TD"}; // royal flush (9, 0)
    //std::string test[5] = {"9D", "KD", "JD", "QD", "TD"}; // straight flush (8, 11)
    //std::string test[5] = {"AD", "AH", "AC", "AS", "5C"}; // four of a kind (7, 12)
    //std::string test[5] = {"9D", "9H", "9C", "8S", "8C"}; // full house (6, 7)
    //std::string test[5] = {"2D", "5D", "JD", "QD", "7D"}; // flush (5, 10)
    //std::string test[5] = {"AD", "KH", "JC", "QS", "TC"}; // royal straight (4, 12)
    //std::string test[5] = {"2D", "3H", "4C", "5S", "AC"}; // five-high straight (4, 3)
    //std::string test[5] = {"7D", "3H", "4C", "5S", "6C"}; // straight (4, 5)
    //std::string test[5] = {"7D", "7H", "7C", "5S", "6C"}; // three (3, 5)
    //std::string test[5] = {"7D", "7H", "6C", "5S", "6H"}; // two (2, 5)
    //std::string test[5] = {"2D", "6H", "6C", "5S", "4H"}; // pair (1, 4)
    //std::string test[5] = {"AD", "6H", "9C", "5S", "4H"}; // high (0, 12)
    //std::array<int, 2> score = evaluate(test);
    //std::cout << score[0] << " " << score[1] << "\n";
    
    //hands = std::vector<std::string>(hands.begin(), hands.begin() + 5);
    std::cout << poker_hands(hands) << "\n";
}
